libraries/AP_RangeFinder/RangeFinder.cpp

```diff
@@ -325,7 +326,18 @@ void RangeFinder::init(enum Rotation orientation_default)
   around 10Hz by main loop
  */
 void RangeFinder::update(void)
-{
+{    
+#if HAL_WITH_UAVCAN
+    if (num_instances < RANGEFINDER_MAX_INSTANCES) {
+        if (drivers[num_instances] == nullptr && params[num_instances].type == RangeFinder_TYPE_UAVCAN) {
+            //keep looking until found
+            drivers[num_instances] = AP_RangeFinder_UAVCAN::detect(state[num_instances], params[num_instances]);
+            if (drivers[num_instances] != nullptr) {
+                num_instances++;
```

Comment:
This doesn't honor the contract of `num_instances` being the highest detected instance correctly, and can corrupt other sensors.

Reply:
While that sounds reasonable, it actually isn't the API we have in a lot of libraries. `num_instances` is the number of the highest detected backend. If you don't setup rangefinder 1, but do setup 2 and detect a driver on 2, then `num_instances` will be 2. https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_RangeFinder/RangeFinder.cpp#L312

The problem here is if someone leaves an empty slot, then things will go wrong.